File: src/com/intellij/idea/plugin/hybris/business/process/diagram/impl/BpDiagramDataModel.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2015 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.business.process.diagram.impl;

import com.intellij.diagram.DiagramDataModel;
import com.intellij.diagram.DiagramNode;
import com.intellij.diagram.DiagramRelationshipInfoAdapter;
import com.intellij.diagram.presentation.DiagramLineType;
import com.intellij.idea.plugin.hybris.business.process.common.BpGraphNode;
import com.intellij.idea.plugin.hybris.business.process.diagram.BpDiagramProvider;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.ModificationTracker;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

/**
 * Created 11:11 PM 31 January 2016.
 *
 * @author Alexander Bartash <AlexanderBartash@gmail.com>
 */
public final class BpDiagramDataModel extends DiagramDataModel<BpGraphNode> {

    private final Collection<BpDiagramFileEdge> edges = new ArrayList<BpDiagramFileEdge>();
    private final Map<String, BpDiagramFileNode> ids2Nodes = new HashMap<String, BpDiagramFileNode>();

    public BpDiagramDataModel(final Project project, final BpGraphNode bpGraphNode) {
        super(project, ServiceManager.getService(BpDiagramProvider.class));

        for (BpGraphNode graphNode : bpGraphNode.getNodesMap().values()) {
            final BpDiagramFileNode fileNode = new BpDiagramFileNode(graphNode);

            this.ids2Nodes.put(graphNode.getGenericAction().getId(), fileNode);
        }
    }

    @NotNull
    @Override
    public Collection<BpDiagramFileNode> getNodes() {
        return this.ids2Nodes.values();
    }

    @NotNull
    @Override
    public Collection<BpDiagramFileEdge> getEdges() {
        return this.edges;
    }

    @NotNull
    @Override
    public String getNodeName(final DiagramNode<BpGraphNode> diagramNode) {
        return diagramNode.getIdentifyingElement().getGenericAction().getId();
    }

    @Contract(value = "_ -> null", pure = true)
    @Nullable
    @Override
    public BpDiagramFileNode addElement(final BpGraphNode t) {
        return null;
    }

    @Override
    public void refreshDataModel() {
        this.edges.clear();

        for (BpDiagramFileNode node : this.ids2Nodes.values()) {
            final BpGraphNode bpGraphNode = node.getIdentifyingElement();

            for (Map.Entry<String, BpGraphNode> transition : bpGraphNode.getTransitions().entrySet()) {
                final String name = transition.getKey();
                final BpGraphNode graphNode = transition.getValue();

                final BpDiagramFileEdge edge = new BpDiagramFileEdge(
                    node, this.ids2Nodes.get(graphNode.getGenericAction().getId()), new BpDiagramRelationship(name)
                );

                this.edges.add(edge);
            }
        }
    }

    @Contract(pure = true)
    @NotNull
    @Override
    public ModificationTracker getModificationTracker() {
        return this;
    }

    @Override
    public void dispose() {

    }

    protected static class BpDiagramRelationship extends DiagramRelationshipInfoAdapter {

        private final String label;

        public BpDiagramRelationship(final String label) {
            super("SOFT", DiagramLineType.SOLID);
            this.label = label;
        }

        @Override
        public Shape getStartArrow() {
            return STANDARD;
        }

        @Override
        public String getLabel() {
            return label;
        }
    }
}

File: src/com/intellij/idea/plugin/hybris/diagram/businessProcess/node/BpDiagramDataModel.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.diagram.businessProcess.node

import com.intellij.diagram.DiagramDataModel
import com.intellij.diagram.DiagramNode
import com.intellij.diagram.DiagramProvider
import com.intellij.idea.plugin.hybris.diagram.businessProcess.BpGraphService
import com.intellij.idea.plugin.hybris.diagram.businessProcess.node.graph.BpGraphFactory
import com.intellij.idea.plugin.hybris.diagram.businessProcess.node.graph.BpGraphNode
import com.intellij.idea.plugin.hybris.diagram.businessProcess.node.graph.BpGraphNodeNavigable
import com.intellij.idea.plugin.hybris.diagram.businessProcess.node.graph.BpGraphNodeRoot
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.ModificationTracker
import org.jetbrains.annotations.Contract
import java.io.Serial

class BpDiagramDataModel(
    project: Project,
    private val rootBpGraphNode: BpGraphNode?,
    provider: DiagramProvider<BpGraphNode?>
) : DiagramDataModel<BpGraphNode?>(project, provider) {

    private val edges: MutableCollection<BpDiagramEdge> = mutableSetOf()
    private val nodesMap: MutableMap<String, BpDiagramNode> = mutableMapOf()

    override fun getNodes() = nodesMap.values
    override fun getEdges() = edges
    override fun getNodeName(diagramNode: DiagramNode<BpGraphNode?>) = diagramNode.identifyingElement.name

    @Contract(value = "_ -> null", pure = true)
    override fun addElement(t: BpGraphNode?): BpDiagramNode? = null

    override fun refreshDataModel() {
        if (rootBpGraphNode !is BpGraphNodeRoot) return

        edges.clear()
        nodesMap.clear()

        val graphService = BpGraphService.getInstance(this.project)
        graphService.buildNodes(rootBpGraphNode)
            .values
            .forEach {
                val bpDiagramFileNode = BpDiagramNode(it, provider)
                nodesMap[it.name] = bpDiagramFileNode
            }

        nodesMap[rootBpGraphNode.name] = BpDiagramNode(rootBpGraphNode, provider)

        nodesMap["Context Parameters"]
            ?.let { BpGraphFactory.buildEdge("parameters", nodesMap[rootBpGraphNode.name]!!, it) }
            ?.let { edges.add(it) }

        nodesMap.values
            .filter { it.identifyingElement is BpGraphNodeNavigable }
            .forEach { targetBpDiagramFileNode ->
                val sourceBpGraphNode = targetBpDiagramFileNode.identifyingElement

                (sourceBpGraphNode as BpGraphNodeNavigable).transitions
                    .forEach { (transitionName, targetBpGraphNode) ->
                        nodesMap[targetBpGraphNode.name]
                            ?.let { sourceBpDiagramFileNode -> BpGraphFactory.buildEdge(transitionName, sourceBpDiagramFileNode, targetBpDiagramFileNode) }
                            ?.let { edges.add(it) }
                    }
            }
    }

    @Contract(pure = true)
    override fun getModificationTracker(): ModificationTracker = this

    override fun dispose() {
        edges.clear()
        nodesMap.clear()
    }

    companion object {
        @Serial
        private const val serialVersionUID: Long = -4815844034119153893L
    }
}
